asllib/carpenter/main.ml

```diff
@@ -0,0 +1,459 @@
+(****************************************************************************************************************)
+(*  SPDX-FileCopyrightText: Copyright 2022-2024 Arm Limited and/or its affiliates <open-source-office@arm.com>  *)
+(*  SPDX-License-Identifier: BSD-3-Clause                                                                       *)
+(****************************************************************************************************************)
+
+open Carpenter_lib
+open Asllib
+
+type 'a target = Itself | BInterp of 'a
+
+let pp_target fmt = function
+  | Itself -> Format.fprintf fmt "no"
+  | BInterp s -> Format.fprintf fmt "BInterp at %s" s
+
+let progress_frequency = 1
+
+let print_progress should_progress =
+  if should_progress then fun i n ->
+    if i mod progress_frequency == 0 then Printf.printf "\r%d/%d%!" i n else ()
```

Comment:
Progress report output  will clobber valid output in case of redirection. Ideally, progress report could be disabled in case of redirection. However, I see no simple solution for now. As a mitigating measure, one can output the progress report on standard error:
```suggestion
    if i mod progress_frequency == 0 then Printf.eprintf "\r%d/%d%!" i n else ()
```

Reply:
Finally I found how to proceed:
```OCaml
let isatty chan =
   Unix.descr_of_out_channel chan |> Unix.isatty

let print_progress should_progress =
  if should_progress && isatty stderr then fun i n -> ...
```

Reply:
As for having progress output on stderr, users may want to redirect the normal output of the tool and still monitor progress. As for `isatty` you are right not to find this very useful, especially since de default behaviour is not outputting the progress reports. But well it does not harm at all for a small benefit...